File: JavaProject/GrowNet/src/main/java/ai/nektron/grownet/InputLayer2D.java
package ai.nektron.grownet;

import java.util.List;

/**
 * Shape‑aware sensory Layer (e.g., grayscale image) using unified onInput/onOutput.
 * This layer creates a single-slot InputNeuron for every pixel in a (height × width) grid.
 */
public class InputLayer2D extends Layer {

    private final int height;
    private final int width;

    /**
     * Manage our own input neurons; no E/I/M neurons are created by the base.
     * @param gain scales each pixel value before thresholding
     * @param epsilonFire tiny floor; values below this do not trigger a spike
     */
    public InputLayer2D(int height, int width, double gain, double epsilonFire) {
        super(/*excitatoryCount*/ 0, /*inhibitoryCount*/ 0, /*modulatoryCount*/ 0);
        this.height = height;
        this.width  = width;

        final List<Neuron> list = getNeurons();
        for (int y = 0; y < height; ++y) {
            for (int x = 0; x < width; ++x) {
                // InputNeuron is single‑slot; share the layer’s bus
                InputNeuron n = new InputNeuron(
                        "IN[" + y + "," + x + "]",
                        getBus(),
                        gain,
                        epsilonFire
                );
                list.add(n);
            }
        }
    }

    /** Row‑major index helper. */
    public int index(int y, int x) { return y * width + x; }

    /**
     * Drive this input Layer with a 2D image (values in [0, 1] or any float range).
     * Follows the same onInput/onOutput contract as other layers.
     */
    public void forwardImage(double[][] image) {
        // Optional: strict shape check
        if (image.length != height || (height > 0 && image[0].length != width)) {
            throw new IllegalArgumentException(
                    "image shape mismatch: expected " + height + "×" + width);
        }

        for (int y = 0; y < height; ++y) {
            for (int x = 0; x < width; ++x) {
                double value = image[y][x];
                InputNeuron n = (InputNeuron) getNeurons().get(index(y, x));
                boolean fired = n.onInput(value);     // unified API: single argument
                if (fired) n.onOutput(value);         // keeps the onOutput contract
            }
        }
    }

    /** Convenience for dashboards: snapshot the last output of each input neuron as an image. */
    public double[][] getFrame() {
        double[][] frame = new double[height][width];
        int idx = 0;
        for (int y = 0; y < height; ++y) {
            for (int x = 0; x < width; ++x) {
                frame[y][x] = ((InputNeuron) getNeurons().get(idx++)).getOutputValue();
            }
        }
        return frame;
    }

    /** Entry layers do not perform intra‑layer routing (no @Override unless Layer defines it). */
    public void propagateFrom(int sourceIndex, double value) {
        // no‑op by design for InputLayer2D
    }
}

File: grownet_python_mojo/python/input_layer_2d.py
from typing import List
from region_bus import RegionBus
from input_neuron import InputNeuron

class InputLayer2D:
    def __init__(self, height: int, width: int, gain: float, epsilon_fire: float, bus: RegionBus) -> None:
        self._h = int(height)
        self._w = int(width)
        self._gain = float(gain)
        self._eps = float(epsilon_fire)
        self._bus = bus
        self._neurons: List[InputNeuron] = [InputNeuron(f"IN_{r}_{c}", bus) for r in range(self._h) for c in range(self._w)]

    def get_neurons(self) -> List[InputNeuron]:
        return self._neurons

    def forward(self, value: float) -> None:
        # broadcast scalar to all inputs
        for n in self._neurons:
            n.on_input(value)

    def forward_image(self, frame: List[List[float]]) -> None:
        # flatten row-major; clip to [h,w]
        for r in range(min(self._h, len(frame))):
            row = frame[r]
            for c in range(min(self._w, len(row))):
                v = float(row[c]) * self._gain
                if abs(v) >= self._eps:
                    self._neurons[r*self._w + c].on_input(v)

    def end_tick(self) -> None:
        self._bus.end_tick()
        for n in self._neurons:
            n.end_tick()
